File: Components/DevExpressVCL/Demos/ExpressSpellChecker/CBuilder/MegaDemo/AddDictionaryForm.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "AddDictionaryForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxButtonEdit"
#pragma link "cxButtons"
#pragma link "cxContainer"
#pragma link "cxControls"
#pragma link "cxDropDownEdit"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxGroupBox"
#pragma link "cxHyperLinkEdit"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxMaskEdit"
#pragma link "cxRadioGroup"
#pragma link "cxTextEdit"
#pragma resource "*.dfm"
TfmAddDictionary *fmAddDictionary;
//---------------------------------------------------------------------------
__fastcall TfmAddDictionary::TfmAddDictionary(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TfmAddDictionary::Add(TdxCustomSpellChecker *ASpellChecker)
{
  TdxSpellCheckerDictionaryItem *ADictionaryItem;
  ADictionaryItem = ASpellChecker->DictionaryItems->Add();
  switch (rgDictionatyType->ItemIndex) {
	case 0: {
	  ADictionaryItem->DictionaryTypeClass = __classid(TdxHunspellDictionary);
	  TdxHunspellDictionary *AHunspellDictionary;
	  AHunspellDictionary = (TdxHunspellDictionary*) ADictionaryItem->DictionaryType;
	  AHunspellDictionary->GrammarPath = beAffFile->Text;
	  AHunspellDictionary->DictionaryPath = beDicFile->Text;
	  AHunspellDictionary->Language = (int) cbLang->Properties->Items->Objects[cbLang->ItemIndex];
	  break;
	}
	case 1: {
	  ADictionaryItem->DictionaryTypeClass = __classid(TdxOpenOfficeDictionary);
	  TdxOpenOfficeDictionary *AOpenOfficeDictionary;
	  AOpenOfficeDictionary = (TdxOpenOfficeDictionary*) ADictionaryItem->DictionaryType;
	  AOpenOfficeDictionary->GrammarPath = beAffFile->Text;
	  AOpenOfficeDictionary->DictionaryPath = beDicFile->Text;
	  AOpenOfficeDictionary->Language = (int) cbLang->Properties->Items->Objects[cbLang->ItemIndex];
	  break;
	}
	case 2: {
	  ADictionaryItem->DictionaryTypeClass = __classid(TdxISpellDictionary);
	  TdxISpellDictionary *AISpellDictionary;
	  AISpellDictionary = (TdxISpellDictionary*) ADictionaryItem->DictionaryType;
	  AISpellDictionary->GrammarPath = beAffFile->Text;
	  AISpellDictionary->DictionaryPath = beDicFile->Text;
	  AISpellDictionary->Language = (int) cbLang->Properties->Items->Objects[cbLang->ItemIndex];
	  break;
	}
  }
  ShowHourglassCursor();
  try
  {
	ADictionaryItem->DictionaryType->Load(dlmDirectLoad);
  }
  __finally
  {
	HideHourglassCursor();
  }
}

//---------------------------------------------------------------------------

void __fastcall TfmAddDictionary::FormCreate(TObject *Sender)
{
  TdxSpellCheckerCodePages *ACodePages;
  int I;
  {
	for (I = 0; I < dxLanguages()->Count; I++) {
	  cbLang->Properties->Items->AddObject(dxLanguages()->Name[I], (TObject*)dxLanguages()->LocaleID[I]);
	}
	cbLang->ItemIndex = dxLanguages()->IndexOf((unsigned)dxLanguages()->GetDefaultLanguageLCID());
  }
  ACodePages = new TdxSpellCheckerCodePages(True);
  try
  {
	for (I = 0; I < ACodePages->Count; I++) {
	  cbCodePage->Properties->Items->AddObject(ACodePages->Name[I], (TObject*)ACodePages->Code[I]);
	}
	for (I = 0; I < ACodePages->Count; I++) {
	  if (ACodePages->Code[I] == GetACP()) {
		cbCodePage->ItemIndex = I; 
		break;
	  }
	}
  }
  __finally
  {
	ACodePages->Free();
  }

}
//---------------------------------------------------------------------------

void __fastcall TfmAddDictionary::rgDictionatyTypePropertiesChange(TObject *Sender)
          
{
  TcxRadioGroup *ARadioGroup = (TcxRadioGroup*)Sender;
  Label4->Visible = ARadioGroup->ItemIndex == 2;
  cbCodePage->Visible = Label4->Visible;
  Label5->Visible = !Label4->Visible;	
  hlLink->Visible = !Label4->Visible;
}
//---------------------------------------------------------------------------

void __fastcall TfmAddDictionary::beAffFilePropertiesButtonClick(TObject *Sender,
          int AButtonIndex)
{
  OpenDialog1->FileName = "";
  OpenDialog1->Filter = "Affix files (*.aff)|*.aff|All files (*.*)|*.*";
  if (OpenDialog1->Execute()) {
	beAffFile->Text = OpenDialog1->FileName;  
  }
}
//---------------------------------------------------------------------------

void __fastcall TfmAddDictionary::beDicFilePropertiesButtonClick(TObject *Sender,
		  int AButtonIndex)
{
  OpenDialog1->FileName = "";
  OpenDialog1->Filter = "Dictionary files (*.dic)|*.dic|All files (*.*)|*.*";
  if (OpenDialog1->Execute()) {
	beDicFile->Text = OpenDialog1->FileName;
  }
}
//---------------------------------------------------------------------------

void __fastcall TfmAddDictionary::beAffFilePropertiesChange(TObject *Sender)
{
  btnAdd->Enabled = FileExists(beAffFile->Text) && FileExists(beDicFile->Text);	
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressSpellChecker/Delphi/MegaDemo/AddDictionaryForm.pas
unit AddDictionaryForm;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, cxGraphics, cxControls, cxLookAndFeels, cxLookAndFeelPainters,
  cxContainer, cxCheckListBox, cxEdit, cxGroupBox, cxRadioGroup,
  cxTextEdit, cxHyperLinkEdit, dxSpellChecker, Menus, StdCtrls, cxButtons,
  ExtCtrls, cxDropDownEdit, cxMaskEdit, cxButtonEdit, cxClasses;

type
  TfmAddDictionary = class(TForm)
    rgDictionatyType: TcxRadioGroup;
    hlLink: TcxHyperLinkEdit;
    btnAdd: TcxButton;
    btnCancel: TcxButton;
    beAffFile: TcxButtonEdit;
    beDicFile: TcxButtonEdit;
    OpenDialog1: TOpenDialog;
    Label1: TLabel;
    Label2: TLabel;
    cbLang: TcxComboBox;
    cbCodePage: TcxComboBox;
    Label3: TLabel;
    Label4: TLabel;
    Bevel1: TBevel;
    Label5: TLabel;
    procedure rgDictionatyTypePropertiesChange(Sender: TObject);
    procedure beAffFilePropertiesButtonClick(Sender: TObject;
      AButtonIndex: Integer);
    procedure beDicFilePropertiesButtonClick(Sender: TObject;
      AButtonIndex: Integer);
    procedure FormCreate(Sender: TObject);
    procedure beAffFilePropertiesChange(Sender: TObject);
  private
  public
    procedure Add(ASpellChecker: TdxCustomSpellChecker);
  end;

var
  fmAddDictionary: TfmAddDictionary;

procedure AddDictionary(ASpellChecker: TdxCustomSpellChecker);

implementation

{$R *.dfm}

uses
  dxISpellDecompressor, dxHunspellDictionary, dxSpellCheckerUtils;

procedure AddDictionary(ASpellChecker: TdxCustomSpellChecker);
begin
  if fmAddDictionary.ShowModal = mrOk then
    fmAddDictionary.Add(ASpellChecker);
end;

{ TfmAddDictionary }

procedure TfmAddDictionary.FormCreate(Sender: TObject);
var
  ACodePages: TdxSpellCheckerCodePages;
  I: Integer;
begin
  for I := 0 to dxLanguages.Count - 1 do
    cbLang.Properties.Items.AddObject(dxLanguages.Name[I], Pointer(dxLanguages.LocaleID[I]));
  cbLang.ItemIndex := dxLanguages.IndexOf(dxLanguages.GetDefaultLanguageLCID);
  ACodePages := TdxSpellCheckerCodePages.Create(True);
  try
    for I := 0 to ACodePages.Count - 1 do
      cbCodePage.Properties.Items.AddObject(ACodePages.Name[I], Pointer(ACodePages.Code[I]));
    for I := 0 to ACodePages.Count - 1 do
      if ACodePages.Code[I] = GetACP then
      begin
        cbCodePage.ItemIndex := I;
        Break;
      end;
  finally
    ACodePages.Free;
  end;
end;

procedure TfmAddDictionary.Add(ASpellChecker: TdxCustomSpellChecker);
const
  DictionaryClasses: array[0..2] of TdxCustomSpellCheckerDictionaryClass = (
    TdxHunspellDictionary, TdxOpenOfficeDictionary, TdxISpellDictionary);
var
  ADictionaryItem: TdxSpellCheckerDictionaryItem;
begin
  ADictionaryItem := ASpellChecker.DictionaryItems.Add;
  ADictionaryItem.DictionaryTypeClass := DictionaryClasses[rgDictionatyType.ItemIndex];
  case rgDictionatyType.ItemIndex of
    0: with TdxHunspellDictionary(ADictionaryItem.DictionaryType) do
       begin
         GrammarPath := beAffFile.Text;
         DictionaryPath := beDicFile.Text;
         Language := Integer(cbLang.Properties.Items.Objects[cbLang.ItemIndex]);
       end;
    1: with TdxOpenOfficeDictionary(ADictionaryItem.DictionaryType) do
       begin
         GrammarPath := beAffFile.Text;
         DictionaryPath := beDicFile.Text;
         Language := Integer(cbLang.Properties.Items.Objects[cbLang.ItemIndex]);
       end;
    2: with TdxISpellDictionary(ADictionaryItem.DictionaryType) do
       begin
         GrammarPath := beAffFile.Text;
         DictionaryPath := beDicFile.Text;
         Language := Integer(cbLang.Properties.Items.Objects[cbLang.ItemIndex]);
         CodePage := Integer(cbCodePage.Properties.Items.Objects[cbCodePage.ItemIndex]);
       end;
  end;
  ShowHourglassCursor;
  try
    ADictionaryItem.DictionaryType.Load(dlmDirectLoad);
  finally
    HideHourglassCursor;
  end;
end;

procedure TfmAddDictionary.rgDictionatyTypePropertiesChange(
  Sender: TObject);
begin
  Label4.Visible := TcxRadioGroup(Sender).ItemIndex = 2;
  cbCodePage.Visible := Label4.Visible;
  Label5.Visible := not Label4.Visible;
  hlLink.Visible := not Label4.Visible;
end;

procedure TfmAddDictionary.beAffFilePropertiesButtonClick(Sender: TObject;
  AButtonIndex: Integer);
begin
  OpenDialog1.FileName := '';
  OpenDialog1.Filter := 'Affix files (*.aff)|*.aff|All files (*.*)|*.*';
  if OpenDialog1.Execute then
    beAffFile.Text := OpenDialog1.FileName;
end;

procedure TfmAddDictionary.beDicFilePropertiesButtonClick(Sender: TObject;
  AButtonIndex: Integer);
begin
  OpenDialog1.FileName := '';
  OpenDialog1.Filter := 'Dictionary files (*.dic)|*.dic|All files (*.*)|*.*';
  if OpenDialog1.Execute then
    beDicFile.Text := OpenDialog1.FileName;
end;

procedure TfmAddDictionary.beAffFilePropertiesChange(Sender: TObject);
begin
  btnAdd.Enabled := FileExists(beAffFile.Text) and FileExists(beDicFile.Text);
end;

end.
